telemetry/metrics.go

```diff
@@ -70,56 +35,91 @@ type GatherResponse struct {
 	ContentType string
 }
 
-// New creates a new instance of Metrics
-func New(cfg Config) (*Metrics, error) {
-	if !cfg.Enabled {
+// ─── Factory ────────────────────────────────────────────────────────────────────
+
+// New creates a new instance of metrics
+func New(opts ...Option) (Metrics, error) {
+	var err error
+
+	c := new(Config)
+	c.loadDefaults()
+	for _, fn := range opts {
+		err = fn(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !c.Enabled {
 		return nil, nil
 	}
 
-	if numGlobalLables := len(cfg.GlobalLabels); numGlobalLables > 0 {
-		parsedGlobalLabels := make([]metrics.Label, numGlobalLables)
-		for i, gl := range cfg.GlobalLabels {
+	if numGlobalLables := len(c.GlobalLabels); numGlobalLables > 0 {
+		parsedGlobalLabels := make([]gometrics.Label, numGlobalLables)
+		for i, gl := range c.GlobalLabels {
 			parsedGlobalLabels[i] = NewLabel(gl[0], gl[1])
 		}
 
-		globalLabels = parsedGlobalLabels
+		c.globalLabels = parsedGlobalLabels
 	}
 
-	metricsConf := metrics.DefaultConfig(cfg.ServiceName)
-	metricsConf.EnableHostname = cfg.EnableHostname
-	metricsConf.EnableHostnameLabel = cfg.EnableHostnameLabel
+	metricsConf := gometrics.DefaultConfig(c.ServiceName)
+	metricsConf.EnableHostname = c.EnableHostname
+	metricsConf.EnableHostnameLabel = c.EnableHostnameLabel
 
-	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
-	metrics.DefaultInmemSignal(memSink)
+	memSink := gometrics.NewInmemSink(10*time.Second, time.Minute)
+	gometrics.DefaultInmemSignal(memSink)
 
-	m := &Metrics{memSink: memSink}
-	fanout := metrics.FanoutSink{memSink}
+	m := &metrics{memSink: memSink, cnf: c}
+	fanout := gometrics.FanoutSink{memSink}
+	var promSink *metricsprom.PrometheusSink
 
-	if cfg.PrometheusRetentionTime > 0 {
-		m.prometheusEnabled = true
-		prometheusOpts := metricsprom.PrometheusOpts{
-			Expiration: time.Duration(cfg.PrometheusRetentionTime) * time.Second,
-		}
+	//Initialize the external library only once.
+	initOnce.Do(func() {
```

Comment:
since this block of code works based on a singleton pattern, running this block of code twice will cause panic. (duplication registration )